finality-aleph/src/data_io.rs

```diff
@@ -1,28 +1,62 @@
 use crate::{Error, Metrics};
 use aleph_bft::OrderedBatch;
-use futures::channel::mpsc;
+use futures::{
+    channel::{mpsc, oneshot},
+    stream::StreamExt,
+};
 use log::debug;
+use parking_lot::Mutex;
 use sp_consensus::SelectChain;
 use sp_runtime::traits::{Block, Header};
+use std::sync::Arc;
 
 #[derive(Clone)]
-pub(crate) struct DataIO<B: Block, SC: SelectChain<B>> {
-    pub(crate) select_chain: SC,
+pub(crate) struct DataIO<B: Block> {
+    pub(crate) best_chain: Arc<Mutex<B::Hash>>,
     pub(crate) ordered_batch_tx: mpsc::UnboundedSender<OrderedBatch<B::Hash>>,
+    pub(crate) refresh_best_chain_tx: mpsc::UnboundedSender<()>,
```

Comment:
If the sole purpose of the channel is to periodically inform the refresh task to request a new best chain, maybe it would be easier to simply have a ticker in the refresh task? Not sure about that though, but the channel seems a bit unintuitive, especially that when we send the request before we read the best chain from the shared variable...

Reply:
I thought about ticker for some time, it has advantage of being simpler, but I thought that the solution with channel has this advantage that the pace of the calls to `best_chain` follows the pace of `get_data`, which depends on AlephBFT and may change. 
I think that refreshing before reading is fine, since we want to get the freshest version.
However, I don't find neither of those two satisfactory, so I you think that ticker is better, I can change it.

Reply:
I don't think the refresh is likely to be in effect when we get the hash, especially since `get_data` is not `async`, so won't give control away. I think with the current semantics of `get_data` (get the newest hash we have at this point) a timer is just better. Unless there is a way to receive push notifications about new blocks being available, but even if that is possible that would be a future improvement rather than a part of this task (unless it's trivial, but I doubt it).

Reply:
Changed to ticker.

Reply:
How about using the long-forgotten bounded channel? We could use `futures::channel::mpsc::channel` which returns a sender which blocks until the previously sent hash is received. And the Receiver has a synchronous `try_next` method which checks whether an item is available. This way the type of `best_chain` can be changed from `Arc<Mutex<B::Hash>>` to `RefCell<B::Hash>`, and `refresh_best_chain` only needs `Sender<B::Hash>` instead of `mpsc::UnboundedReceiver<()>` and `Arc<Mutex<B::Hash>>`.

Reply:
I would say making this mechanism better should be a separate task, since I suspect there might be a much better approach with push notifications.

Besides, I think what @michalseweryn proposes might have the problem of giving us the next block, rather than the newest? But I'm not sure I understood this correctly, so I may be way off.

Reply:
I'll add a task to Jira, to come up with optimal solution and let stick to the ticker for now. Is it ok with you?

Reply:
sure